agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,191 @@
+package hcp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/hashstructure/v2"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+	"github.com/hashicorp/consul/agent/hcp/telemetry"
+)
+
+var (
+	// internalMetricRefreshFailure is a metric to monitor refresh failures.
+	internalMetricRefreshFailure []string = []string{"hcp", "telemetry_config_provider", "refresh", "failure"}
+	// internalMetricRefreshFailure is a metric to monitor refresh successes.
+	internalMetricRefreshSuccess []string = []string{"hcp", "telemetry_config_provider", "refresh", "success"}
+)
+
+// Ensure hcpProviderImpl implements telemetry provider interfaces.
+var _ telemetry.ConfigProvider = &hcpProviderImpl{}
+var _ telemetry.EndpointProvider = &hcpProviderImpl{}
+
+// hcpProviderImpl holds telemetry configuration and settings for continuous fetch of new config from HCP.
+// it updates configuration, if changes are detected.
+type hcpProviderImpl struct {
+	// cfg holds configuration that can be dynamically updated.
+	cfg *dynamicConfig
+	// cfgHash is used to check if two dynamicConfig objects are equal.
+	cfgHash uint64
+
+	// A reader-writer mutex is used as the provider is read heavy.
+	// OTEL components access telemetryConfig during metrics collection and export (read).
+	// Meanwhile, config is only updated when there are changes (write).
+	rw sync.RWMutex
+	// hcpClient is an authenticated client used to make HTTP requests to HCP.
+	hcpClient client.Client
+}
+
+// dynamicConfig is a set of configurable settings for metrics collection, processing and export.
+type dynamicConfig struct {
+	endpoint *url.URL
+	labels   map[string]string
+	filters  *regexp.Regexp
+	// refreshInterval controls the interval at which configuration is fetched from HCP to refresh config.
+	refreshInterval time.Duration
+}
+
+// providerParams is used to initialize a hcpProviderImpl.
+type providerParams struct {
+	metricsConfig   *client.MetricsConfig
+	refreshInterval time.Duration
+	hcpClient       client.Client
+}
+
+// NewHCPProviderImpl initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProviderImpl(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
```

Comment:
```suggestion
func NewHCPProvider(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
```

Personally I wouldn't include `Impl` in the public func. We know it's an implementation of the interface, but it's the only one